lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -686,27 +687,39 @@ SILCombiner::createApplyWithConcreteType(FullApplySite AI,
   }
   Args.push_back(NewSelf);
 
-  // Form a new set of substitutions where Self is
-  // replaced by a concrete type.
-  SmallVector<Substitution, 8> Substitutions;
-  for (auto Subst : AI.getSubstitutions()) {
-    auto *A = Subst.getReplacement()->getAs<ArchetypeType>();
-    if (A && A == OpenedArchetype) {
-      auto Conformances = AI.getModule().getASTContext()
-                            .AllocateUninitialized<ProtocolConformanceRef>(1);
-      Conformances[0] = Conformance;
-      Substitution NewSubst(ConcreteType, Conformances);
-      Substitutions.push_back(NewSubst);
-    } else
-      Substitutions.push_back(Subst);
-  }
-
+  auto FnTy = AI.getCallee()->getType().castTo<SILFunctionType>();
   SILType SubstCalleeType = AI.getSubstCalleeSILType();
-
   SILType NewSubstCalleeType;
 
-  auto FnTy = AI.getCallee()->getType().castTo<SILFunctionType>();
+  // Form a new set of substitutions where Self is
+  // replaced by a concrete type.
+  SmallVector<Substitution, 8> Substitutions;
   if (FnTy->isPolymorphic()) {
+    auto FnSubsMap =
+        FnTy->getGenericSignature()->getSubstitutionMap(AI.getSubstitutions());
+    auto FinalSubsMap = FnSubsMap.subst(
+        [&](SubstitutableType *type) -> Type {
+          if (type == OpenedArchetype)
+            return ConcreteType;
+          return type;
+        },
+        [&](CanType origTy, Type substTy,
+            ProtocolType *proto) -> Optional<ProtocolConformanceRef> {
+          if (substTy->getCanonicalType() == ConcreteType) {
+            if (proto->getDecl() != Conformance.getRequirement()) {
+              assert(
+                  Conformance.getRequirement()->inheritsFrom(proto->getDecl()));
+              if (Conformance.isAbstract())
+                return ProtocolConformanceRef(proto->getDecl());
+              Conformance = ProtocolConformanceRef(
+                  Conformance.getConcrete()->getInheritedConformance(
+                      proto->getDecl()));
```

Comment:
This bit and the similar snippet on line 813 could be factors out into a ProtocolConformanceRef::getInheritedConformance